contracts/src/protocol/OmniPortal.sol

```diff
@@ -158,40 +163,60 @@ contract OmniPortal is IOmniPortal, IOmniPortalAdmin, OwnableUpgradeable, OmniPo
      *          - xmsg().sender         == address(0)
      */
     function xmsg() external view returns (XTypes.MsgShort memory) {
-        return _currentXmsg;
+        return _xmsg;
     }
 
     /**
      * @notice Returns true the current transaction is an xcall, false otherwise
      */
     function isXCall() external view returns (bool) {
-        return _currentXmsg.sourceChainId != 0;
+        return _xmsg.sourceChainId != 0;
     }
 
     /**
      * @notice Execute an xmsg.
      * @dev Verify an XMsg is next in its XStream, execute it, increment inXStreamOffset, emit an XReceipt
      */
     function _exec(XTypes.Msg calldata xmsg_) internal {
-        require(xmsg_.destChainId == chainId, "OmniPortal: wrong destChainId");
+        require(
+            xmsg_.destChainId == chainId || xmsg_.destChainId == _BROADCAST_CHAIN_ID, "OmniPortal: wrong destChainId"
+        );
         require(xmsg_.streamOffset == inXStreamOffset[xmsg_.sourceChainId] + 1, "OmniPortal: wrong streamOffset");
 
         // set xmsg to the one we're executing
-        _currentXmsg = XTypes.MsgShort(xmsg_.sourceChainId, xmsg_.sender);
+        _xmsg = XTypes.MsgShort(xmsg_.sourceChainId, xmsg_.sender);
 
         // increment offset before executing xcall, to avoid reentrancy loop
         inXStreamOffset[xmsg_.sourceChainId] += 1;
 
-        // we enforce a maximum on xcall, but we trim to max here just in case
-        uint256 gasLimit = xmsg_.gasLimit > XMSG_MAX_GAS_LIMIT ? XMSG_MAX_GAS_LIMIT : xmsg_.gasLimit;
+        uint256 gasUsed;
+        bool success;
+
+        if (xmsg_.to == _VIRUTAL_PORTAL_ADDRESS) {
```

Comment:
seeing some duplicated logic here, can we avoid the large if/else blocks? e.g. with ternary operators

Reply:
I did that initially, but honestly thought this was clearer. A little duplication doesn't hurt.

Plus gaslimit gets awkward when you try to share too much logic